clang/lib/CConv/TypeVariableAnalysis.cpp

```diff
@@ -0,0 +1,185 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Classes dealing with the processing of type variables for generic functions.
+//===----------------------------------------------------------------------===//
+
+#include "clang/CConv/TypeVariableAnalysis.h"
+
+using namespace llvm;
+using namespace clang;
+
+std::set<ConstraintVariable *> &TypeVariableEntry::getConstraintVariables() {
+  assert("Accessing ConstraintVariable set for inconsistent Type Variable." &&
+      IsConsistent);
+  return ArgConsVars;
+}
+
+void TypeVariableEntry::insertConstraintVariables
+    (std::set<ConstraintVariable *> &CVs) {
+  assert("Accessing ConstraintVariable set for inconsistent Type Variable." &&
+      IsConsistent);
+  ArgConsVars.insert(CVs.begin(), CVs.end());
+}
+
+void TypeVariableEntry::setTypeParamConsVar(ConstraintVariable *CV) {
+  assert("Accessing constraint variable for inconsistent Type Variable." &&
+      IsConsistent);
+  assert("Setting constraint variable to null" && CV != nullptr);
+  assert("Changing already set constraint variable" &&
+      TypeParamConsVar == nullptr);
+  TypeParamConsVar = CV;
+}
+
+void TypeVariableEntry::updateEntry(QualType Ty,
+                                    std::set<ConstraintVariable *> &CVs) {
+  QualType PointeeType = Ty->getPointeeType();
+  if (isTypeAnonymous(PointeeType)) {
+    // We'll need a name to provide the type arguments during rewriting, so
+    // no anonymous things here.
+    IsConsistent = false;
+  } else if (IsConsistent && getType() != Ty) {
+    // If it has previously been instantiated as a different type, its use
+    // is not consistent.
+    IsConsistent = false;
+  } else if (IsConsistent) {
+    // Type variable has been encountered before with the same type. Insert
+    // new constraint variables.
+    insertConstraintVariables(CVs);
+  }
+  // If none of the above branches are hit, then this was already inconsistent,
+  // so there's no need to update anything.
+}
+
+ConstraintVariable *TypeVariableEntry::getTypeParamConsVar() {
+  assert("Accessing constraint variable for inconsistent Type Variable." &&
+      IsConsistent);
+  assert("Accessing null constraint variable" && TypeParamConsVar != nullptr);
+  return TypeParamConsVar;
+}
+
+QualType TypeVariableEntry::getType() {
+  assert("Accessing Type for inconsistent Type Variable." &&
+      IsConsistent);
+  return TyVarType;
+}
+
+bool TypeVariableEntry::getIsConsistent() const {
+  return IsConsistent;
+}
+
+// Finds cast expression that contain function call to a generic function. If
+// the return type of the function uses a type variable, a binding for the
+// return is added to the type variable map.
+bool TypeVarVisitor::VisitCastExpr(CastExpr *CE) {
```

Comment:
It might be that the cast is not consistent with what variable you will be inserting. Are you checking that here?

Reply:
TypeVariableEntry::updateEntry handles consistency checks.

Reply:
This is what I mean: Suppose you have
```
int **p = (int **) malloc(sizeof(int *)*10);
```
3C might convert this to 
```
_Array_ptr<_Ptr<int>> p = (int **)malloc<_Ptr<int>>(sizeof(int *)*10);
```
but the Checked C compiler will complain because the `int **` cast is not allowed. What I am asking is whether you could realize that this is going to happen, and stop type instantiation altogether by not rewriting `p`.

Reply:
I think the solution to this is the same as  #160. We can remove the `(int **)`, rewrite it to `(_Ptr<_Ptr<int>>)`, or keep `p` as `int **`.

Reply:
I was thinking about what should happen in the meantime. I guess we can ignore it if you are working on #160